src/transformers/modeling_reformer.py

```diff
@@ -0,0 +1,1757 @@
+            weight.expand((batch_size,) + self.axial_pos_shape + weight.shape[-1:]) for weight in self.weights
+        ]
+
+        if self.training is True:
+            assert (
+                reduce(mul, self.axial_pos_shape) == sequence_length
+            ), "Make sure that config.axial_pos_shape factors: {} multiply to sequence length: {}".format(
+                self.axial_pos_shape, sequence_length
+            )
+            if self.dropout > 0:
+                weights = torch.cat(broadcasted_weights, dim=-1)
+                # permute weights so that 2D correctly drops dims 1 and 2
+                perm_weigthts = weights.permute(0, 3, 2, 1)
+                # drop entire matrix of last two dims (prev dims 1 and 2)
+                drop_perm_weights = nn.functional.dropout2d(perm_weigthts, self.dropout, training=self.training)
+                drop_weights = drop_perm_weights.permute(0, 3, 2, 1)
+
+                position_encodings = torch.reshape(drop_weights, (batch_size, sequence_length, -1))
+
+            else:
+                position_encodings = torch.cat(
+                    [torch.reshape(weight, (batch_size, sequence_length, -1)) for weight in broadcasted_weights],
+                    dim=-1,
+                )
+
+        else:
+            assert (
+                reduce(mul, self.axial_pos_shape) >= sequence_length
+            ), "Make sure that config.axial_pos_shape factors: {} multiply at least to max(sequence_length, least_common_mult_chunk_length): max({}, {})".format(
+                self.axial_pos_shape, sequence_length, self.least_common_mult_chunk_length,
+            )
+
+            # reshape axial encodings and use only until sequence_length
+            position_encodings = torch.cat(broadcasted_weights, dim=-1)
+            position_encodings = position_encodings.view(batch_size, -1, position_encodings.shape[-1])[
+                :, :sequence_length
+            ]
+
+        return position_encodings
+
+
+class PositionEmbeddings(nn.Module):
+    """Constructs conventional position embeddings of shape `[max_pos_embeddings, hidden_size]`.
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        self.dropout = config.hidden_dropout_prob
+        self.embedding = nn.Embedding(config.max_position_embeddings, config.hidden_size)
+
+    def forward(self, position_ids):
+        position_embeddings = self.embedding(position_ids)
+        position_embeddings = nn.functional.dropout(position_embeddings, self.dropout, self.training)
+        return position_embeddings
+
+
+class ReformerEmbeddings(nn.Module):
+    """Construct the embeddings from word, position and token_type embeddings.
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        self.max_position_embeddings = config.max_position_embeddings
+        self.dropout = config.hidden_dropout_prob
+
+        self.word_embeddings = nn.Embedding(config.vocab_size, config.hidden_size)
```

Comment:
padding_idx?

Reply:
this is called `embed_tokens` in other models IIRC

Reply:
Some models do ``nn.Embedding(..., padding_idx=config.padding_idx)`` others don't ... what is the general rule here? @thomwolf ? Is it really necessary to add the padding_idx to the ``nn.Embedding`` ? What do we do if there is no padding idx?